### DISCUSSION TITLE
Nothing is being rendered

### DISCUSSION THREAD

**Discussion by helloimalastair** [maplibre/maplibre-gl-js#2191] [Q&A] [answered]
Hey y'all!

Trying to get my new tile server started, and I've hit a weird snag. The tileserver seems to be responding with tiles correctly, but when MapLibre attempts to render it, all it generates is a tan background. I can see it making the requests correctly, and it isn't logging any errors, so I'm kind of stumped on what the issue might be...

I initially thought it had something to do with me choosing invalid coordinates, since I only generated tiles for Denmark, but whether I set it to coordinates within Denmark, or just plain `0/0/0`, it still doesn't render anything.

Anyone have any ideas what I might be missing?

[Demo](https://validconventionaltranslation.helloimalastair.repl.co/)
[Code for Demo](https://replit.com/@HelloImAlastair/MapLibre-Demo)

**Comment by HarelM** [comment]
maplibre doesn't support `mapbox://` protocol.
I haven't looked too deeply into the code, but I would advise to start of with the examples and the tiles used there and complicate things as you get familiar with the library and stack.

**Reply by helloimalastair** [reply]
Rereading my code after a few days, do you think it might have something to do with the fact that I don't have an API key for Maptiler? I've been looking for how to host `glyphs` on my own, but still not sure how...

On the topic of the `mapbox://` link, wasn't that just in the metadata? I would assume that wouldn't have any effect on the actual rendering, so I removed it.

**Comment by zerda-ocm** [comment]
The tiles your tileserver serves, appear to be empty:
https://observablehq.com/@henrythasler/mapbox-vector-tile-dissector

Your issue seems to be the tile generation.

**Reply by helloimalastair** [reply]
It's really odd. I was able to get it to generate a rough outline of the map before, with just the landmass shown, but now nothing loads at al...

**Reply by zerda-ocm** [reply]
How do you generate your tiles (OpenMapTiles, planetiler, etc)? And what tileserver do you use?

**Reply by helloimalastair** [reply]
I'm using [Protomaps' PMTiles](https://github.com/protomaps/PMTiles/) system, with a Cloudflare [Workers](https://workers.cloudflare.com)/[R2](https://developers.cloudflare.com/r2) Tileserver.

To generate the `.pmtiles` file, I download an `.osm.pbf` file from *Geofabrik*, run it through [`tilemaker`](https://github.com/systemed/tilemaker) to convert to `.mbtiles`, and then through the [`pmtiles`](https://github.com/protomaps/go-pmtiles) CLI.

**Reply by zerda-ocm** [reply]
You can use the PMTiles Viewer to see what data your pmtiles contain. 
https://protomaps.github.io/PMTiles/

Is there a reason why you use a tileserver? The benefit of pmtiles is, that there is no need for a tileserver:
https://github.com/protomaps/PMTiles/blob/main/js/examples/maplibre.html

**Reply by helloimalastair** [reply]
This is the output on the PMTiles viewer:
![image](https://user-images.githubusercontent.com/37487746/219952600-f6eaaee9-3d00-42f1-a9d1-6791dab6a3f5.png)

By Tileserver, I just meant their [Cloudflare Adapter](https://github.com/protomaps/PMTiles/tree/main/serverless/cloudflare), which takes the coordinate format in the URL(z.x.y.mvt), and turns it into `Range` Requests in R2. It is only a tile server in the sense that it "serves tiles", but it doesn't actually generate them on-the-spot.

**Comment by zerda-ocm** [comment]
I don't have any experience with the Cloudfare Adapter, but I am using https://github.com/protomaps/PMTiles/blob/main/js/examples/maplibre.html without problems.
(has the benefit you don't have to pay for Cloudflare Workers)

**Reply by helloimalastair** [reply]
Weird... I'm still getting nothing rendered, even though I can see the 206 Partial Content Responses in the Network Tab...

[Demo](https://maps.goalastair.com/)

If it matters, it is still running through a Worker, but this time, it is just a straight proxy to the R2 custom domain:
```js
const html = `HTML_STRING`; // Cut for brevity
export default {
  async fetch(req) {
    const url = new URL(req.url);
    if(url.pathname === "/") {
      return new Response(html, {
        headers: {
          "content-type": "text/html"
        }
      });
    }
    let res = await fetch(req);
    res = new Response(res.body, res);
    res.headers.set("access-control-allow-origin", "*");
    res.headers.set("access-control-allow-methods", "GET, OPTIONS");
    return res;
  }
}
```

**Comment by zerda-ocm** [comment, marked as answer]
see attached file for a simple example that works with your provided pmfile:
[test_goalastair.zip](https://github.com/maplibre/maplibre-gl-js/files/10777249/test_goalastair.zip)

**Reply by helloimalastair** [reply]
Ok, seems to be working, at least a bit. [Demo](https://maps.goalastair.com/). Though, is there a way to get it to render the different surfaces/street names, rather than just basic shapes?

**Reply by zerda-ocm** [reply]
https://github.com/openmaptiles/maptiler-basic-gl-style/blob/master/style.json is a good starting point.

Street names for example are defined in the layer "road_major_label"

Regarding [surfaces](https://wiki.openstreetmap.org/wiki/Key:surface): You have to update your tilemaker process to have this information included in your pmtiles.

**Reply by helloimalastair** [reply]
~~Just to clarify, does [this config](https://github.com/systemed/tilemaker/blob/master/resources/config-openmaptiles.json) not cover the required Surface generation?~~
Looks like it does generate **most** of the layers required, but weirdly, only some of the bodies of water are present. Is that because I only selected the map for Denmark? Because it is also cutting out some of the bodies of water within Denmark's borders: [Demo](https://maps.goalastair.com)

Also, for the glyphs/sprites in the `style` config, is there a good place where I can find some resources for self-hosting these?

**Reply by zerda-ocm** [reply]
Could be a missing file in your tilemaker setup: "coastline/water_polygons.shp", but I haven't used tilemaker and therefore don't really know.

glyphs:
https://github.com/openmaptiles/fonts
https://github.com/maplibre/font-maker

**Reply by helloimalastair** [reply]
Thanks anyway! Hope you have a nice week!
